### PR TITLE
Fix for declarative pipelines sometimes dropping downstream build links

### PR DESCRIPTION
**Pull request by sophistifunk**
# Description

Actions were not being correctly collected from the parallel branches when using declarative pipeline and not wrapping a branch for labelling / sequential-stage purposes. This caused the UI to not show downstream builds under these circumstances.

Test pipeline for demonstration (requires some pipelines name `downstream1` and `downstream2`):

```
pipeline {
    agent any
    stages {
        stage ("Stage the first") {
            steps {
                echo "This pipeline waits for the downstream builds to finish"
            }
        }
        stage ("Stage the second") {
            parallel {
                stage ("downstream1") {
                    steps {
                        build 'downstream1'
                    }
                }
                stage ("downstream2") {
                    steps {
                        build 'downstream2'
                    }
                }
            }
        }
        stage ("Double-downstream") {
            steps {
                echo 'start ds1, with wait'
                build 'downstream1'
                echo 'start ds2, with wait'
                build 'downstream2'
            }
        }
    }
}
```

See [JENKINS-38339](https://issues.jenkins-ci.org/browse/JENKINS-38339).

# Submitter checklist
- [X] Link to JIRA ticket in description, if appropriate.
- [X] Change is code complete and matches issue description
- [X] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [X] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/FlowNodeWrapper.java [hunk 1]
```diff
@@ -221,6 +222,14 @@ public void setBlockErrorAction(ErrorAction blockErrorAction) {
         return filtered;
     }
 
+    /**
+     * Returns Action instances that were attached to the associated FlowNode, or to any of its children
+     * not represented in the graph.
+     */
+    public Collection<Action> getPipelineActions() {
```
Comment (sophistifunk) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/FlowNodeWrapper.java]: Hadn't needed this before now.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java [hunk 2]
```diff
@@ -378,49 +388,74 @@ public void parallelStart(@Nonnull FlowNode parallelStartNode, @Nonnull FlowNode
...
                 if (stack.isEmpty()) {
+                    // We've got a single non-nested stage for this branch. We're going to discard it from our graph,
+                    // since we use the node for the branch itself, which has been created with the same name...
+
+                    if (isNodeVisitorDumpEnabled) {
+                        dump("\t\tSingle-stage branch");
+                    }
+
+                    // ...but first we need to poach any actions from the stage node we'll be discarding
+                    branchActions.addAll(firstNodeWrapper.getPipelineActions());
```
Comment (sophistifunk) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: This line is the meat of the fix. The rest is logic cleanup for readability and documentation, and slightly better info when debug mode is on.

File: blueocean-pipeline-api-impl/src/main/java/org/jenkinsci/plugins/workflow/job/views/GraphDumpAction.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
+/*
```
Comment (sophistifunk) [inline blueocean-pipeline-api-impl/src/main/java/org/jenkinsci/plugins/workflow/job/views/GraphDumpAction.java]: This is a debugging helper, built using GraphVizAction as a template. I'm leaving it here for now, but I'll talk to Pipeline see if they want it.

File: blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/downstreamBuildLinksDecl.jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,32 @@
+pipeline {
```
Comment (sophistifunk) [inline blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/downstreamBuildLinksDecl.jenkinsfile]: Same as `downstreamBuildLinks.jenkinsfile` but declarative.

File: blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/downstreamBuildLinksSeq.jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,47 @@
+pipeline {
```
Comment (sophistifunk) [inline blueocean-pipeline-api-impl/src/test/resources/io/jenkins/blueocean/rest/impl/pipeline/downstreamBuildLinksSeq.jenkinsfile]: Mixes simple branch, single-stage named branch, and sequential stage branch for peace of mind.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java [hunk 6]
```diff
@@ -311,18 +317,18 @@ public void parallelStart(@Nonnull FlowNode parallelStartNode, @Nonnull FlowNode
                                branchNode.getDisplayName(), branchNode.getDisplayFunctionName()));
         }
 
-        if ( nestedbranches.size() != parallelBranchEndNodes.size() ) {
+        if (nestedbranches.size() != parallelBranchEndNodes.size()) {
             logger.debug(String.format("nestedBranches size: %s not equal to parallelBranchEndNodes: %s",
-                nestedbranches.size(), parallelBranchEndNodes.size()));
-            if(!parallelEnds.isEmpty()){
+                                       nestedbranches.size(), parallelBranchEndNodes.size()));
```
Comment (NicuPascu) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: the indentation looks a tad off here

Comment (sophistifunk) [reply blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: Nah, those are args to `String.format()` not `logger.debug()`

**Comment by ctataryn** [comment]
Really looking forward to this PR being merged. Good job, thank you! Just wanted to be sure, this will allow us to click a link in the log that indicates the launching of a downstream and have that link open the log for that corresponding downstream job?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java [hunk 7]
```diff
@@ -378,49 +388,74 @@ public void parallelStart(@Nonnull FlowNode parallelStartNode, @Nonnull FlowNode
...
+                        dump("\t\tSingle-stage branch");
+                    }
+
+                    // ...but first we need to poach any actions from the stage node we'll be discarding
+                    branchActions.addAll(firstNodeWrapper.getPipelineActions());
+
+                } else {
+                    // We've got nested stages for sequential stage branches and/or branch labelling purposes
+
+                    // Usually ignore firstNodeWrapper, but if the fisrt stage has a different name...
```
Comment (halkeye) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeGraphVisitor.java]: nit: fisrt => first

File: blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeTest.java [hunk 8]
```diff
@@ -3067,8 +3065,103 @@ public void downstreamBuildLinks() throws Exception {
         assertThat("node #4 contains a link to downstream1", actions, hasItem(actionMatcher2));
     }
 
+    @Test
+    @Issue("JENKINS-38339")
+    public void downstreamBuildLinksDeclarative() throws Exception {
```
Comment (halkeye) [inline blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeTest.java]: image

TESTSSSSSSSSSSS

**Comment by sophistifunk** [comment]
@ctataryn it doesn't change the downstream link functionality at all, just fixes some cases in which it wasn't being triggered.

**Pull request merged**